fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java

```diff
@@ -0,0 +1,152 @@
+import org.apache.doris.nereids.trees.expressions.Expression;
+import org.apache.doris.nereids.trees.expressions.NamedExpression;
+import org.apache.doris.nereids.trees.expressions.Slot;
+import org.apache.doris.nereids.trees.expressions.SlotReference;
+import org.apache.doris.nereids.trees.expressions.functions.AggregateFunction;
+import org.apache.doris.nereids.trees.plans.GroupPlan;
+import org.apache.doris.nereids.trees.plans.Plan;
+import org.apache.doris.nereids.trees.plans.logical.LogicalAggregate;
+import org.apache.doris.nereids.trees.plans.logical.LogicalFilter;
+import org.apache.doris.nereids.trees.plans.logical.LogicalHaving;
+
+import com.google.common.collect.Maps;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Resolve having clause to the aggregation.
+ */
+public class ResolveHaving extends OneAnalysisRuleFactory {
+    @Override
+    public Rule build() {
+        return RuleType.RESOLVE_HAVING.build(
+            logicalHaving(logicalAggregate()).thenApply(ctx -> {
+                LogicalHaving<LogicalAggregate<GroupPlan>> having = ctx.root;
+                LogicalAggregate<GroupPlan> aggregate = having.child();
+                Set<Expression> groupByExpressions = new HashSet<>(aggregate.getGroupByExpressions());
+                Set<Expression> tokens = extractTokensForAggregation(having.getPredicates())
+                        .collect(Collectors.toSet());
+                tokens.forEach(token -> {
+                    if (token instanceof SlotReference && (!groupByExpressions.contains(token)
+                            && aggregate.getOutputExpressions().stream()
+                            .noneMatch(output -> {
+                                if (output instanceof SlotReference) {
+                                    return output.equals(token);
+                                } else if (output instanceof Alias) {
+                                    Alias alias = (Alias) output;
+                                    return alias.toSlot().equals(token) || alias.child().equals(token);
+                                }
+                                return false;
+                            }))) {
+                        throw new AnalysisException(token.toSql() + " in having clause should be grouped by.");
+                    } else if (token instanceof AggregateFunction
+                            && checkWhetherNestedAggregateFunctionsExist((AggregateFunction) token)) {
+                        throw new AnalysisException("Aggregate functions in having clause can't be nested: "
+                                + token.toSql() + ".");
+                    }
+                });
+                return createLogicalFilter(having, tokens);
```

Comment:
if we put some expression to child's output expression, we need add a project on filter to keep the output constant